src/compiler/checker.ts

```diff
@@ -18434,24 +18440,28 @@ namespace ts {
 
         function checkDeclarationInitializer(declaration: VariableLikeDeclaration) {
             const type = getTypeOfExpression(declaration.initializer, /*cache*/ true);
+            return shouldUseLiteralType(declaration) ? type : getWidenedLiteralType(type);
+        }
+
+        function shouldUseLiteralType(declaration: VariableLikeDeclaration) {
             return getCombinedNodeFlags(declaration) & NodeFlags.Const ||
                 getCombinedModifierFlags(declaration) & ModifierFlags.Readonly && !isParameterPropertyDeclaration(declaration) ||
-                isTypeAssertion(declaration.initializer) ? type : getWidenedLiteralType(type);
+                (declaration.initializer && isTypeAssertion(declaration.initializer));
         }
 
-        function isLiteralContextualType(contextualType: Type) {
+        function isLiteralContextualType(contextualType: Type, candidateLiteral: Type): boolean {

```

Comment:
Consider flipping the parameters and changing the name to canBeContextuallyTypedAsLiteral (or something similar; probably the parameter names would need to change too.)

Reply:
@rbuckton 's dynamic names PR already includes both of those things, actually.